### PR TITLE
Add OMEMO wire protocol data classes

### PR DESCRIPTION
**Pull request by melvo**

### CODE REVIEW

**Comment by lnjX** [comment]
Can you rebase upon https://github.com/qxmpp-project/qxmpp/commit/00e2b4f857b099f08ed2c8bf4da234c67e6ec694?

**Comment by melvo** [comment]
@lnj I rebased it.

**Comment by lnjX** [comment]
This is your error message:
https://github.com/qxmpp-project/qxmpp/pull/349/checks?check_run_id=2641970105#step:6:565

File: src/base/QXmppOmemoDeviceListElement.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+/// The ID must be a positive value, otherwise it is ignored.
+///
+/// \param id this device list element's ID
+///
+void QXmppOmemoDeviceListElement::setId(const int id)
+{
+    if (id > 0) {
+        d->id = id;
+    }
+}
```
Comment (lnjX) [inline src/base/QXmppOmemoDeviceListElement.cpp:98]: You can't reset the ID to its initial state (0)? And maybe you want to use an unsigned integer.

File: src/base/QXmppOmemoDeviceListElement.cpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ */
+
```
Comment (lnjX) [inline src/base/QXmppOmemoDeviceListElement.cpp:24]: can you put the implementation of the objects into one file? (e.g. `QXmppOmemoData.cpp`) the header file can stay as they are

this makes the compilation faster (and in some cases the result is better optimized, but probably not here)

File: src/base/QXmppOmemoDeviceList.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+/// \param other
+///
+QXmppOmemoDeviceList &QXmppOmemoDeviceList::operator=(const QXmppOmemoDeviceList &other) = default;
+
+/// \cond
+void QXmppOmemoDeviceList::parse(const QDomElement &element)
+{
+    QDomElement device = element.firstChildElement("device");
+
+    while (!device.isNull()) {
```
Comment (lnjX) [inline src/base/QXmppOmemoDeviceList.cpp:68]: for loop plz

File: src/base/QXmppOmemoDeviceBundle.cpp [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+///
+/// The key of a key-value pair represents the ID of the corresponding public pre key.
+/// The ID must be a positive value, otherwise the corresponding key-value pair is ignored.
+///
+/// \param keys public pre keys
+///
+void QXmppOmemoDeviceBundle::setPublicPreKeys(const QMap<int, QByteArray> &keys)
+{
+    for (auto it = keys.cbegin(); it != keys.cend(); it++) {
+        if (it.key() > 0) {
```
Comment (lnjX) [inline src/base/QXmppOmemoDeviceBundle.cpp:188]: use a `uint` here too

File: src/base/QXmppOmemoEnvelope.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ */
+
+#ifndef QXMPPOMEMOENVELOPE_H
+#define QXMPPOMEMOENVELOPE_H
+
+#include "QXmppElement.h"
```
Comment (lnjX) [inline src/base/QXmppOmemoEnvelope.h:28]: remove

File: tests/qxmppomemodata/tst_qxmppomemodata.cpp [hunk 6]
```diff
@@ -25,6 +25,7 @@
 #include "QXmppOmemoDeviceBundle.h"
 #include "QXmppOmemoDeviceList.h"
 #include "QXmppOmemoDeviceListElement.h"
+#include <QXmppOmemoEnvelope.h>
```
Comment (lnjX) [inline tests/qxmppomemodata/tst_qxmppomemodata.cpp:29]: `<>` -> `""`

File: src/base/QXmppOmemoElement.cpp [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+///
+/// Sets the ID of the sender's device.
+///
+/// The ID must be a positive value, otherwise it is ignored.
+///
+/// \param id sender's device ID
+///
+void QXmppOmemoElement::setSenderDeviceId(const int id)
```
Comment (lnjX) [inline src/base/QXmppOmemoElement.cpp:92]:
```
 void QXmppOmemoElement::setSenderDeviceId(int id)
```

File: src/base/QXmppOmemoElement.cpp [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+/// Searches for an OMEMO envelope by its recipient JID and device ID.
+///
+/// \param recipientJid bare JID of the recipient
+/// \param recipientDeviceId ID of the recipient's device
+///
+/// \return the found OMEMO envelope
+///
+std::optional<QXmppOmemoEnvelope> QXmppOmemoElement::searchEnvelope(const QString &recipientJid, int recipientDeviceId) const
+{
+    for (const auto &envelope : d->envelopes.values(recipientJid)) {
```
Comment (lnjX) [inline src/base/QXmppOmemoElement.cpp:129]: * detaches `d->envelopes.values(recipientJid)`
* better use a linear search here, that's still better than creating a new QList and copying elements

File: src/base/QXmppOmemoElement.cpp [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+/// \cond
+void QXmppOmemoElement::parse(const QDomElement &element)
+{
+    const auto header = element.firstChildElement("header");
+
+    d->senderDeviceId = header.attribute("sid").toInt();
+
+    auto recipient = header.firstChildElement("keys");
+    while (!recipient.isNull()) {
```
Comment (lnjX) [inline src/base/QXmppOmemoElement.cpp:161]: for loop

File: src/base/QXmppOmemoElement.cpp [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+        const auto recipientJid = recipient.attribute("jid");
+        auto envelope = recipient.firstChildElement("key");
+
+        while (!envelope.isNull()) {
```
Comment (lnjX) [inline src/base/QXmppOmemoElement.cpp:165]: for loop

File: src/base/QXmppOmemoElement.cpp [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+
+    writer->writeStartElement("header");
+    writer->writeAttribute("sid", QString::number(d->senderDeviceId));
+
+    const auto recipientJids = d->envelopes.uniqueKeys();
+    for (const auto &recipientJid : recipientJids) {
+        writer->writeStartElement("keys");
+        writer->writeAttribute("jid", recipientJid);
+
+        for (const auto &envelope : d->envelopes.values(recipientJid)) {
```
Comment (lnjX) [inline src/base/QXmppOmemoElement.cpp:192]: same problem with .values() here too

**Comment by lnjX** [comment]
Merged: https://github.com/qxmpp-project/qxmpp/commit/0623aa38f2ead734dddea4cbad899a868f01cb1e (and others)

**Pull request closed without merging**